etcd.go

```diff
@@ -0,0 +1,127 @@
+package dsync
+
+import (
+	"github.com/coreos/etcd/client"
+	"github.com/pborman/uuid"
+	"golang.org/x/net/context"
+	"sync"
+	"time"
+)
+
+type etcdMutex struct {
+	backoff BackoffFunc
+	ctx     context.Context
+	etcd    client.KeysAPI
+	key     string
+	locked  bool
+	m       sync.Mutex
+	refresh time.Duration
+	stop    chan struct{}
+	ttl     time.Duration
+	uuid    string
+	wg      sync.WaitGroup
+}
+
+func (e *etcdMutex) Lock() {
+	var i int64
+	e.m.Lock()
+	for e.locked != false {
+		// The lock is already held within this process.
+		// Wait for the lock to be unlocked.
+	}
+	e.stop = make(chan struct{})
+	// Error if the previous value of the key is not empty.
+	var options = client.SetOptions{PrevExist: client.PrevNoExist, TTL: e.ttl}
+	for i = 0; e.locked != true; i++ {
+		select {
+		case <-e.ctx.Done():
+			// The parent context may decide to Timeout or Cancel,
+			// in that case, stop trying to obtain the lock.
+			return
+		default:
+			_, err := e.etcd.Set(e.ctx, e.key, e.uuid, &options)
+			if err == nil {
+				e.locked = true
+				break
+			}
+			if i < 0 { // Safeguard against potential overflows.
+				i = 0
+			}
+			time.Sleep(e.backoff(i))
+		}
+	}
+	go e.keepAlive() // Keep the lock alive by refreshing the TTL.
+	e.m.Unlock()
+}
+
+func (e *etcdMutex) keepAlive() {
+	e.wg.Add(1)
```

Comment:
Can multiple goroutines call these methods at the same time. In some places you are using a lock and not in others?